gcloud/storage/batch.py

```diff
@@ -0,0 +1,208 @@
+            headers['Content-Type'] = 'application/json'
+            headers['Content-Length'] = len(body)
+        if body is None:
+            body = ''
+        lines = ['%s %s HTTP/1.1' % (method, uri)]
+        lines.extend(['%s: %s' % (key, value)
+                      for key, value in sorted(headers.items())])
+        lines.append('')
+        lines.append(body)
+        payload = '\r\n'.join(lines)
+        if sys.version_info[0] < 3:  # pragma: NO COVER  Python2
+            MIMEApplication.__init__(self, payload, 'http', encode_noop)
+        else:                        # pragma: NO COVER  Python3
+            super_init = super(MIMEApplicationHTTP, self).__init__
+            super_init(payload, 'http', encode_noop)
+
+
+class Batch(object):
+    """Proxy an underlying connection, batching up change operations.
+
+    :type connection: :class:`gcloud.storage.connection.Connection`
+    :param connection: the connection for which the batch proxies.
+    """
+    def __init__(self, connection):
+        self._connection = connection
+        self._http = _FauxHTTP(connection)
+        self._requests = self._responses = ()
+        for attr in _PROXIED_ATTRS:
+            setattr(self, attr, getattr(connection, attr))
+
+    def finish(self):
+        """Submit a single `multipart/mixed` request w/ deferred requests.
+
+        :rtype: list of tuples
+        :returns: one ``(status, reason, payload)`` tuple per deferred request.
+        :raises: ValueError if no requests have been deferred.
+        """
+        deferred = self._requests = self._http.finalize()
+
+        if len(deferred) == 0:
+            raise ValueError("No deferred requests")
+
+        multi = MIMEMultipart()
+
+        for method, uri, headers, body in deferred:
+            sub = MIMEApplicationHTTP(method, uri, headers, body)
+            multi.attach(sub)
+
+        # Flatten payload
+        if six.PY3:             # pragma: NO COVER  Python3
+            buf = io.StringIO()
+        else:                   # pragma: NO COVER  Python2
+            buf = io.BytesIO()
+        generator = Generator(buf, False, 0)
+        generator.flatten(multi)
+        payload = buf.getvalue()
+
+        # Strip off redundant header text
+        _, body = payload.split('\n\n', 1)
+        headers = dict(multi._headers)
+
+        url = self._connection.build_api_url('/batch')
+
+        _req = self._connection._make_request
+        response, content = _req('POST', url, data=payload, headers=headers)
+        self._responses = list(_crack_mime_response(response, content))
+        return self._responses
+
+    def __enter__(self):
+        _BATCHES.push(self)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        try:
+            if exc_type is None:
+                self.finish()
+            else:
+                self._http.reset()
+        finally:
+            _BATCHES.pop()
+
+
+def _crack_mime_response(response, content):
```

Comment:
Can you change. It wasn't clear to me what "crack" meant even after reading the first few lines.